hyperbolic-rs/rudders/losses.py

```diff
@@ -14,133 +14,271 @@
 """Loss functions for CF with support for optional negative sampling."""
 
 import abc
+
 import tensorflow as tf
+from rudders.models import Relations
 
 
 class LossFunction(abc.ABC):
     """Abstract loss function for CF embeddings."""
 
-    def __init__(self, n_users, n_items, args, **kwargs):
-        """Initialize CF loss function.
-
-        Args:
-          sizes: Tuple of size 2 containing (n_users, n_items).
-          neg_sample_size: Integer indicating the number of negative samples to use.
-          double_neg: Bool indicating whether or not to use double negative sampling.
-          margin: Float indicating the margin between a score for positive and negative examples.
-        """
-        self.n_users = n_users
-        self.n_items = n_items
-        self.neg_sample_size = args.neg_sample_size
-        self.margin = tf.Variable(args.margin * tf.keras.backend.ones(1), trainable=False)
-
     @abc.abstractmethod
     def calculate_loss(self, model, input_batch):
         """
-        Args:
-          model: tf.keras.Model CF embedding model.
-          input_batch: Tensor of size batch_size x 2 containing input pairs.
-
-        Returns:
-          Average loss within the input_batch.
+        :param model: CF embedding model.
+        :param input_batch: Tensor of size batch_size x 2 containing input pairs: (user, item)
+        :return: Average loss within the input_batch.
         """
         pass
 
 
-class PairwiseHingeLoss(LossFunction):
-    """Pairwise ranking hinge loss."""
-
-    def __init__(self, n_users, n_items, args, **kwargs):
-        super(PairwiseHingeLoss, self).__init__(n_users, n_items, args)
-        self.gamma = tf.Variable(tf.keras.backend.ones(1), trainable=False)
+class TripletWithNegativeSampleLoss(LossFunction, abc.ABC):
```

Comment:
Cool, I think I get the organization and logic of this.
But as discussed, since this is obsolete at this point I'd rather see the new version.